design/application-abi.md

```diff
@@ -29,12 +36,22 @@ These kinds are mutually exclusive; implementations should report an error if
 asked to instantiate a module containing exports which declare it to be of
 multiple kinds.
 
-Regardless of the kind, all programs accessing WASI APIs also export a linear
-memory with the name `memory`. Pointers in WASI API calls are relative to this
-memory's index space.
+Regardless of the kind, all modules accessing WASI APIs also export a linear
+memory with the name `memory`. Data pointers in WASI API calls are relative to
+this memory's index space.
+
+Regardless of the kind, all modules accessing WASI APIs also export a table
+with the name `__indirect_function_table`. Function pointers in WASI API calls
+are relative to this table's index space.
+
+For compatibility with existing toolchains, modules may also export globals
+named `__heap_base` and `__data_end`. Environments shall not access them.
+This provision is deprecated and toolchains are encouraged to avoid providing
+these exports.
```

Comment:
I'm not sure why this last paragraph is needed?  Did we have some toolchain version where those symbols were exported by default?    Presumably wasi modules are free export any number of additional things on top of what wasi specifies?    If I'm wrong and additional exports are not permitted then we should explicitly state that here.

Reply:
Yes; they're exported by default in the Rust toolchain, due to a series of coincidences. I've now added additional wording clarifying this.